pkg/server/api_v2_grants.go

```diff
@@ -0,0 +1,365 @@
+		// If ok is false, the query returned 0 rows.
+		return resp, nil
+	}
+
+	scanner := makeResultScanner(it.Types())
+	for ; ok; ok, err = it.Next(ctx) {
+		if err != nil {
+			return resp, err
+		}
+		grant := grantRecord{}
+		row := it.Cur()
+		if resp.PaginationInfo.TotalResults == 0 {
+			resp.PaginationInfo.TotalResults = int64(tree.MustBeDInt(row[2]))
+		}
+		if err := scanner.Scan(row, "grantee", &grant.Grantee); err != nil {
+			return resp, err
+		}
+		if err := scanner.Scan(row, "privilege_type", &grant.Privilege); err != nil {
+			return resp, err
+		}
+		resp.Results = append(resp.Results, grant)
+	}
+	return resp, nil
+}
+
+// getTableGrants returns a paginated response of table grants metadata.
+// If the user does not have the required permissions to access the table, we will return a 403 Forbidden error.
+//
+// ---
+// parameters:
+//   - name: tableId
+//     type: integer
+//     description: The ID of the table to get grants for.
+//     in: path
+//
+// produces:
+// - application/json: tableGrantsResponse
+//
+// responses:
+//
+//	"200":
+//		description: A paginated response of grants on the provided table.
+//	"404":
+//		description: The table does not exist.
+//	"400":
+//		description: The request is malformed.
+//	"500":
+//		description: An internal server error occurred.
+func (a *apiV2Server) getTableGrants(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	ctx = a.sqlServer.AnnotateCtx(ctx)
+	sqlUser := authserver.UserFromHTTPAuthInfoContext(ctx)
+	if r.Method != http.MethodGet {
+		http.Error(w, "Method Not Allowed", http.StatusMethodNotAllowed)
+		return
+	}
+
+	// Validate request parameters.
+	pathVars := mux.Vars(r)
+	tableId, err := strconv.Atoi(pathVars[tableIdPathVar])
+	if err != nil || tableId < 0 {
+		http.Error(w, "invalid database id", http.StatusBadRequest)
+		return
+	}
+	queryValues := r.URL.Query()
+	pageSize, err := apiutil.GetIntQueryStringVal(queryValues, pageSizeKey)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	pageNum, err := apiutil.GetIntQueryStringVal(queryValues, pageNumKey)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	offset := 0
+	if pageNum > 0 {
+		offset = (pageNum - 1) * pageSize
+	}
+
+	// Get the table grants.
+	resp, err := getTableGrantsResponseWithPagination(ctx, a.sqlServer.internalExecutor, tableId, sqlUser, pageSize, offset)
+	if err != nil {
+		srverrors.APIV2InternalError(ctx, err, w)
+		return
+	}
+	if resp.TableName == "" {
+		http.Error(w, "table does not exist", http.StatusNotFound)
+		return
+	}
+
+	resp.PaginationInfo.PageNum = pageNum
+	resp.PaginationInfo.PageSize = pageSize
+
+	apiutil.WriteJSONResponse(ctx, w, http.StatusOK, resp)
+}
+
+type tableGrantsResponseWithPagination struct {
+	PaginatedResponse[[]grantRecord]
+	DatabaseName string `json:"database_name"`
+	SchemaName   string `json:"schema_name"`
+	TableName    string `json:"table_name"`
```

Comment:
nit: Why do we need this data? I think the UI would use the response from `GET api/v2/table_metadata/{table_id}/` to get this information.

Reply:
I think this additional metadata doesn't hurt - any reason we shouldn't include it?